File: sort/cycle_sort/Java/CycleSort.java
import java.util.Arrays;

/**
 * Created by aubdira on 10/30/17. CycleSort
 */
public class CycleSort {

    public static void main(String[] args) {
        int[] arr = {233, 51, 63, 7, 18, 30, 66, -5, 128, 65, 49, 4};

        System.out.println(Arrays.toString(arr));

        cycleSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    private static void cycleSort(int[] a) {

        for (int cycleStart = 0; cycleStart < a.length - 1; cycleStart++) {
            int val = a[cycleStart];

            // count the number of values that are smaller than val
            // since cycleStart
            int pos = cycleStart;
            for (int i = cycleStart + 1; i < a.length; i++) {
                if (a[i] < val) {
                    pos++;
                }
            }

            // if there aren't any
            if (pos == cycleStart) {
                continue;
            }

            // skip duplicates
            while (val == a[pos]) {
                pos++;
            }

            // put val into final position
            int tmp = a[pos];
            a[pos] = val;
            val = tmp;

            // repeat as long as we can find values to swap
            // otherwise start new cycle
            while (pos != cycleStart) {
                pos = cycleStart;
                for (int i = cycleStart + 1; i < a.length; i++) {
                    if (a[i] < val) {
                        pos++;
                    }
                }

                while (val == a[pos]) {
                    pos++;
                }

                tmp = a[pos];
                a[pos] = val;
                val = tmp;
            }
        }
    }
}

File: sort/cycle_sort/ruby/cycle_sort.rb
def cycleSort!(array)
  writes = 0
 
  # Loop through the array to find cycles to rotate.
  for cycleStart in 0 .. array.size-2
    item = array[cycleStart]
 
    # Find where to put the item.
    pos = cycleStart
    for i in cycleStart+1 ... array.size
      pos += 1  if array[i] < item
    end
 
    # If the item is already there, this is not a cycle.
    next  if pos == cycleStart
 
    # Otherwise, put the item there or right after any duplicates.
    pos += 1  while item == array[pos]
    array[pos], item = item, array[pos]
    writes += 1
 
    # Rotate the rest of the cycle.
    while pos != cycleStart
 
      # Find where to put the item.
      pos = cycleStart
      for i in cycleStart+1 ... array.size
        pos += 1  if array[i] < item
      end
 
      # Put the item there or right after any duplicates.
      pos += 1  while item == array[pos]
      array[pos], item = item, array[pos]
      writes += 1
    end
  end
  writes
end 
 
p a = [0, 1, 2, 2, 2, 2, 1, 9, 3.5, 5, 8, 4, 7, 0, 6]
puts "writes : #{cycleSort!(a)}"
p a